File: impls/pm-c/step1_read_print.c
#include <stdio.h>

char* read(char* input) {
  return input;
}

char* eval(char* input) {
  return input;
}

char* print(char* input) {
  return input;
}

char* rep(char* input) {
  char* buffer = read(input);
  buffer = eval(buffer);
  buffer = print(buffer);
  return buffer;
}

int main(void) {
  char* line = NULL;
  size_t size;
  printf("user> ");
  while(getline (&line, &size, stdin) != -1) {
    char* result = rep(line);
    printf("%s", result);
    printf("user> ");
  }
  return 0;
}

File: impls/pm-js/step1_read_print.js
const readline = require('readline');
let r = require('./reader')
let p = require('./printer')

function rep(input) {
    output = READ(input);
    output = EVAL(output);
    output = PRINT(output);
    return output;
}

function READ(input) {
    let form = r.read_str(input);
    return form;
}

function EVAL(input) {
    return input;
}

function PRINT(input) {
    return (p.pr_str(input));
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: 'user> '
});

rl.prompt();

rl.on('line', (line) => {
    console.log(rep(line))
    r.Reader.position = 0;
    rl.prompt();
}).on('close', () => {
    console.log("\n");
    process.exit(0);
});
